### ISSUE TITLE
A failure of yarn did not have an error shell return code

### ISSUE THREAD

**Issue by victornoel** [yarnpkg/yarn#2698]
**Do you want to request a *feature* or report a *bug*?**

Bug

**What is the current behavior?**

On my CI system, the following happened (not sure why, but I think it is coming from my VMs, not yarn)
```
$ time yarn run prod-silent && mv dist ../frontend-dist
yarn run v0.20.0
$ yarn run build-silent -- -bh='{{baseUrl}}' && yarn run precache-silent 
yarn run v0.20.0
$ yarn run build -- --progress=false -bh={{baseUrl}}
yarn run v0.20.0
$ ng build --prod --aot --no-sourcemap --progress=false -bh={{baseUrl}}
magicString.insertLeft(...) is deprecated. Use magicString.appendLeft(...) instead
FATAL ERROR: Committing semi space failed. Allocation failed - process out of memory

<--- Last few GCs --->

[536:0x2b18ef0]  1917343 ms: Scavenge 672.9 (723.0) -> 672.5 (724.0) MB, 2465.7 / 193.3 ms  (+ 275.8 ms in 1909 steps since last GC) allocation failure 
[536:0x2b18ef0]  1923425 ms: Scavenge 673.6 (724.0) -> 672.7 (724.0) MB, 1987.3 / 194.0 ms  (+ 827.8 ms in 1923 steps since last GC) allocation failure 
[536:0x2b18ef0]  1931051 ms: Scavenge 673.8 (724.0) -> 672.8 (724.0) MB, 1863.6 / 138.5 ms  (+ 1274.2 ms in 1938 steps since last GC) allocation failure 


<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 0x116a413c0d39 <JS Object>
    2: _serializeMappings(aka SourceMapGenerator_serializeMappings) [/builds/linagora/petals-cockpit/frontend/node_modules/source-map/lib/source-map-generator.js:~291] [pc=0x1484497cfe13](this=0x2598eee7b2f9 <a SourceMapGenerator with map 0x19af89845129>)
    3: toJSON(aka SourceMapGenerator_toJSON) [/builds/linagora/petals-cockpit/frontend/node_modules/source-map/lib/source-map-generator.js:38...

 1: node::Abort() [@angular/cli]
 2: 0x126426c [@angular/cli]
 3: v8::Utils::ReportOOMFailure(char const*, bool) [@angular/cli]
 4: v8::internal::V8::FatalProcessOutOfMemory(char const*, bool) [@angular/cli]
 5: v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::GCCallbackFlags) [@angular/cli]
 6: v8::internal::Heap::CollectGarbage(v8::internal::GarbageCollector, char const*, char const*, v8::GCCallbackFlags) [@angular/cli]
 7: v8::internal::Factory::NewFillerObject(int, bool, v8::internal::AllocationSpace) [@angular/cli]
 8: v8::internal::Runtime_AllocateInTargetSpace(int, v8::internal::Object**, v8::internal::Isolate*) [@angular/cli]
 9: 0x148447708506
Aborted (core dumped)
error Command failed with exit code 134.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.

real	34m12.069s
user	1m46.864s
sys	0m30.128s
```

And instead of failing, the build continued, so I suppose that the shell return code was 0 at that point.

**If the current behavior is a bug, please provide the steps to reproduce.**

Not sure how…

**What is the expected behavior?**

The yarn process should return 1 and not 0 as the shell return code in case of errors like this.

**Please mention your node.js, yarn and operating system version.**

yarn 0.20.0, linux and node 7.5.0

**Comment by bestander** [comment]
Last time I checked Node 7 was crashing quite often because it uses latest and less tested V8

**Comment by victornoel** [comment]
@bestander could be the reason why it crashes yes… still, yarn shouldn't return a return code of 0 in that case :)

**Comment by skirmess** [comment]
We have the same problem. This output is from Yarn 0.18.1 but 0.20.3 has the same issue. We've reduced the old space size to ensure that webpack fails.

```
$ grep compile package.json
    "precompile": "rimraf ./dist",
    "compile": "node --max_old_space_size=10 node_modules/.bin/webpack --config build/webpack.prod.config.js",
$ yarn compile
yarn compile v0.18.1
$ rimraf ./dist
$ node --max_old_space_size=10 node_modules/.bin/webpack --config build/webpack.prod.config.js

<--- Last few GCs --->

     286 ms: Mark-sweep 7.9 (43.9) -> 6.9 (44.9) MB, 4.3 / 0 ms [last resort gc].
     297 ms: Mark-sweep 7.9 (44.9) -> 7.4 (44.9) MB, 4.3 / 0 ms [allocation failure] [GC in old space requested].
     303 ms: Mark-sweep 7.7 (44.9) -> 7.7 (44.9) MB, 4.8 / 0 ms [allocation failure] [GC in old space requested].
     307 ms: Mark-sweep 7.7 (44.9) -> 7.7 (44.9) MB, 4.2 / 0 ms [last resort gc].
     311 ms: Mark-sweep 7.7 (44.9) -> 7.7 (44.9) MB, 3.8 / 0 ms [last resort gc].


<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 0x20779eb4629 <JS Object>
    2: /* anonymous */(aka /* anonymous */) [vm.js:~52] [pc=0x12c4b9fd1904] (this=0x20779e041b9 <undefined>,code=0xfe66903b029 <Very long string[79882]>,options=0x1c88e17c74c9 <an Object with map 0x2a9f768102e1>)
    3: _compile [module.js:373] [pc=0x12c4b9f5cda1] (this=0x1c88e17aaff1 <a Module with map 0x336a95516e79>,content=0xfe669014081 <Very long string[79816]>,filename=0x1c88e17aaf31 <Stri...

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - process out of memory
Done in 0.73s.
```

The problem is that yarn exits with a RC of 0 after that FATAL ERROR.
```
$ echo $?
0
```

Now, let's run it with a wrapper shell script

```
 $ cat xxx
#!/bin/bash

"$@"

$ grep '"compile"' package.json
    "precompile": "rimraf ./dist",
    "compile": "./xxx node --max_old_space_size=10 node_modules/.bin/webpack --config build/webpack.prod.config.js",
$ yarn compile
yarn compile v0.18.1
$ rimraf ./dist
$ ./xxx node --max_old_space_size=10 node_modules/.bin/webpack --config build/webpack.prod.config.js

<--- Last few GCs --->

     645 ms: Mark-sweep 7.9 (43.9) -> 6.9 (44.9) MB, 9.7 / 0 ms [last resort gc].
     670 ms: Mark-sweep 7.9 (44.9) -> 7.4 (44.9) MB, 9.9 / 0 ms [allocation failure] [GC in old space requested].
     680 ms: Mark-sweep 7.7 (44.9) -> 7.7 (44.9) MB, 8.8 / 0 ms [allocation failure] [GC in old space requested].
     688 ms: Mark-sweep 7.7 (44.9) -> 7.7 (44.9) MB, 9.0 / 0 ms [last resort gc].
     697 ms: Mark-sweep 7.7 (44.9) -> 7.7 (44.9) MB, 8.8 / 0 ms [last resort gc].


<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 0x7d9dbab4629 <JS Object>
    2: /* anonymous */(aka /* anonymous */) [vm.js:~52] [pc=0x6b0cdd1f04] (this=0x7d9dba041b9 <undefined>,code=0x2a986b03b029 <Very long string[79882]>,options=0x1ab76e1caa29 <an Object with map 0x237a0dc102e1>)
    3: _compile [module.js:373] [pc=0x6b0cd5cda1] (this=0x5f7767ff059 <a Module with map 0x123ec6916e79>,content=0x2a986b014081 <Very long string[79816]>,filename=0x5f7767fef99 <String[8...

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - process out of memory
./xxx: line 3: 17028 Aborted                 (core dumped) "$@"
error Command failed with exit code 134.
```

This time Yarn recognizes the error and returns a non-zero RC.

```
$ echo $?
1
```

**Comment by bestander** [comment]
Looks like Yarn catches exception when the failing code is wrapped in a shell script then.
The code that handles shells out to scripts is here https://github.com/yarnpkg/yarn/blob/master/src/util/execute-lifecycle-script.js#L191

Any ideas why exception is not thrown?

**Comment by skirmess** [comment]
We get a return code of 134, which means the child process got killed by signal 6 (134-128).

Signal 6 is SIGABRT on my machine.

```
$ kill -l | grep SIGABRT
 6) SIGABRT      7) SIGBUS       8) SIGFPE       9) SIGKILL     10) SIGUSR1
```

strace from the first run where yarn called node directly:

```
29799 write(2, "FATAL ERROR: CALL_AND_RETRY_LAST"..., 75) = 75
29799 rt_sigprocmask(SIG_UNBLOCK, [ABRT], NULL, 8) = 0
29799 tgkill(29799, 29799, SIGABRT)     = 0
29799 --- SIGABRT (Aborted) @ 0 (0) ---
```

strace from the second run where yarn called bash which called node:

```
37247 write(2, "FATAL ERROR: CALL_AND_RETRY_LAST"..., 75) = 75
37247 rt_sigprocmask(SIG_UNBLOCK, [ABRT], NULL, 8) = 0
37247 tgkill(37247, 37247, SIGABRT)     = 0
37247 --- SIGABRT (Aborted) @ 0 (0) ---
37246 <... wait4 resumed> [{WIFSIGNALED(s) && WTERMSIG(s) == SIGABRT && WCOREDUMP(s)}], 0, NULL) = 37247
37246 open("/usr/share/locale/locale.alias", O_RDONLY) = 3
37246 fstat(3, {st_mode=S_IFREG|0644, st_size=2512, ...}) = 0
37246 mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f3b11796000
37246 read(3, "# Locale name alias data base.\n#"..., 4096) = 2512
37246 read(3, "", 4096)                 = 0
37246 close(3)                          = 0
37246 munmap(0x7f3b11796000, 4096)      = 0
37246 open("/usr/share/locale/en_US.UTF-8/LC_MESSAGES/bash.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
37246 open("/usr/share/locale/en_US.utf8/LC_MESSAGES/bash.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
37246 open("/usr/share/locale/en_US/LC_MESSAGES/bash.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
37246 open("/usr/share/locale/en.UTF-8/LC_MESSAGES/bash.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
37246 open("/usr/share/locale/en.utf8/LC_MESSAGES/bash.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
37246 open("/usr/share/locale/en/LC_MESSAGES/bash.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
37246 fstat(2, {st_mode=S_IFCHR|0666, st_rdev=makedev(5, 0), ...}) = 0
37246 ioctl(2, SNDCTL_TMR_TIMEBASE or TCGETS, {B38400 opost isig icanon echo ...}) = 0
37246 mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f3b11796000
37246 open("/usr/share/locale/en_US.UTF-8/LC_MESSAGES/libc.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
37246 open("/usr/share/locale/en_US.utf8/LC_MESSAGES/libc.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
37246 open("/usr/share/locale/en_US/LC_MESSAGES/libc.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
37246 open("/usr/share/locale/en.UTF-8/LC_MESSAGES/libc.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
37246 open("/usr/share/locale/en.utf8/LC_MESSAGES/libc.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
37246 open("/usr/share/locale/en/LC_MESSAGES/libc.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
37246 write(2, "./xxx: line 3: 37247 Aborted    "..., 64) = 64
37246 rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
37246 --- SIGCHLD (Child exited) @ 0 (0) ---
37246 wait4(-1, 0x7fff1f0bf49c, WNOHANG, NULL) = -1 ECHILD (No child processes)
37246 rt_sigreturn(0xffffffffffffffff)  = 0
37246 rt_sigaction(SIGINT, {SIG_DFL, [], SA_RESTORER, 0x3b59a32920}, {0x43d170, [], SA_RESTORER, 0x3b59a32920}, 8) = 0
37246 rt_sigprocmask(SIG_BLOCK, NULL, [], 8) = 0
37246 read(255, "\n", 19)               = 1
37246 rt_sigprocmask(SIG_BLOCK, NULL, [], 8) = 0
37246 read(255, "", 19)                 = 0
37246 exit_group(134)                   = ?
```

The shell returns a non-zero return code after node died. I assume Yarn must somehow catch that in the call to spawn().

**Comment by terribleplan** [comment]
We are observing this during a docker container build...

```
<--- Last few GCs --->

 1359028 ms: Mark-sweep 1965.1 (1406.2) -> 1965.1 (1406.2) MB, 3031.0 / 0.0 ms (+ 0.8 ms in 1 steps since start of marking, biggest step 0.8 ms) [allocation failure] [GC in old space requested].
 1361618 ms: Mark-sweep 1965.1 (1406.2) -> 1973.8 (1403.2) MB, 2589.1 / 0.0 ms [last resort gc].
 1364157 ms: Mark-sweep 1973.8 (1403.2) -> 1982.6 (1403.2) MB, 2539.1 / 0.0 ms [last resort gc].


<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 0x3f190afcfb39 <JS Object>
    1: runInContext [vm.js:~29] [pc=0x23161b563dca] (this=0x2ec2b8d4b1b9 <a ContextifyScript with map 0x1fe9ceb0f9e9>,contextifiedSandbox=0x106e756dd909 <a Window with map 0x17015cab7069>,options=0x3f190af04381 <undefined>)
    2: arguments adaptor frame: 1->2
    3: runScript [/usr/src/app/node_modules/jest-environment-jsdom/build/index.js:~53] [pc=0x23161ac933a8] (this=0x106e756dd949 <a JSDOM...

[91mFATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
[0m[91m 1: [0m[91mnode::Abort() [node]
 2: [0m[91m0x109b9cc [node]
 3: [0m[91mv8::Utils::ReportApiFailure(char const*, char const*) [node]
 4: [0m[91mv8::internal::V8::FatalProcessOutOfMemory(char const*, bool) [node]
 5: [0m[91mv8::internal::Factory::CopyFixedArrayAndGrow(v8::internal::Handle<v8::internal::FixedArray>, int, v8::internal::PretenureFlag) [node][0m[91m
 6: [0m[91mv8::internal::SharedFunctionInfo::AddToOptimizedCodeMap(v8::internal::Handle<v8::internal::SharedFunctionInfo>, v8::internal::Handle<v8::internal::Context>, v8::internal::MaybeHandle<v8::internal::Code>, v8::internal::Handle<v8::internal::LiteralsArray>, v8::internal::BailoutId) [node]
 7: v8::internal::Compiler::PostInstantiation(v8::internal::Handle<v8::internal::JSFunction>, v8::internal::PretenureFlag) [node]
 8: v8::internal::Factory::NewFunctionFromSharedFunctionInfo(v8::internal::Handle<v8::internal::SharedFunctionInfo>, v8::internal::Handle<v8::internal::Context>, v8::internal::PretenureFlag) [node]
 9: [0m[91mv8::UnboundScript::BindToCurrentContext() [node]
10: [0m[91mnode::ContextifyScript::EvalMachine(node::Environment*, long, bool, bool, v8::FunctionCallbackInfo<v8::Value> const&, v8::TryCatch*) [node]
11: node::ContextifyScript::RunInContext(v8::FunctionCallbackInfo<v8::Value> const&) [node]
12: v8::internal::FunctionCallbackArguments::Call(void (*)(v8::FunctionCallbackInfo<v8::Value> const&)) [node]
13: [0m[91m0x9ec36e [node]
14: [0m[91m0x9ecc0e [node]
15: 0x2316150092a7
[0mDone in 1364.57s.
yarn install v0.24.4
[1/4] Resolving packages...
[2/4] Fetching packages...
[91mwarning fsevents@1.0.17: The platform "linux" is incompatible with this module.
[0minfo "fsevents@1.0.17" is an optional dependency and failed compatibility check. Excluding it from installation.
[3/4] Linking dependencies...
[4/4] Building fresh packages...
$ yarn build:dll
yarn build:dll v0.20.3
$ better-npm-run build:dll 
running better-npm-run in /usr/src/app
```

This is as a `RUN` step inside our Dockerfile with commands such as `RUN yarn && yarn build && rm -rf node_modules && yarn --production `.

**Comment by kevinpumpup** [comment]
I am using `yarn` to run test suite (mocha) and a mocha "heap out of memory" error (code 134) is being swallowed by yarn, resulting in false-positive return code from test/CI.

**Comment by BYK** [comment]
The original issue should be fixed via #3995. @kevinpumpup if you are still having issues with the latest version of Yarn (1.2.1) please file a new issue with more details about repro steps.
